extension/jmxmetricsextension/subprocess/subprocess.go

```diff
@@ -0,0 +1,310 @@
+// limitations under the License.
+
+package subprocess
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// exported to be used by jmx metrics extension
+type Config struct {
+	ExecutablePath       string            `mapstructure:"executable_path"`
+	Args                 []string          `mapstructure:"args"`
+	EnvironmentVariables map[string]string `mapstructure:"environment_variables"`
+	StdInContents        string            `mapstructure:"stdin_contents"`
+	RestartOnError       bool              `mapstructure:"restart_on_error"`
+	RestartDelay         *time.Duration    `mapstructure:"restart_delay"`
+	ShutdownTimeout      *time.Duration    `mapstructure:"shutdown_timeout"`
+}
+
+// exported to be used by jmx metrics extension
+type Subprocess struct {
+	Stdout         chan string
+	cancel         context.CancelFunc
+	config         *Config
+	envVars        []string
+	logger         *zap.Logger
+	pid            *pid
+	shutdownSignal chan struct{}
+	// configurable for testing purposes
+	sendToStdIn func(string, io.Writer) error
+}
+
+type pid struct {
+	pid     int
+	pidLock sync.Mutex
+}
+
+func (p *pid) setPid(pid int) {
+	p.pidLock.Lock()
+	defer p.pidLock.Unlock()
+	p.pid = pid
+}
+
+func (p *pid) getPid() int {
+	p.pidLock.Lock()
+	defer p.pidLock.Unlock()
+	return p.pid
+}
+
+func (subprocess *Subprocess) Pid() int {
+	return subprocess.pid.getPid()
+}
+
+// exported to be used by jmx metrics extension
+func NewSubprocess(conf *Config, logger *zap.Logger) *Subprocess {
+	if conf.RestartDelay == nil {
+		defaultRestartDelay := 5 * time.Second
+		conf.RestartDelay = &defaultRestartDelay
+	}
+	if conf.ShutdownTimeout == nil {
+		defaultShutdownTimeout := 5 * time.Second
+		conf.ShutdownTimeout = &defaultShutdownTimeout
+	}
+
+	return &Subprocess{
+		Stdout:         make(chan string),
+		pid:            &pid{pid: -1, pidLock: sync.Mutex{}},
+		config:         conf,
+		logger:         logger,
+		shutdownSignal: make(chan struct{}),
+		sendToStdIn:    sendToStdIn,
+	}
+}
+
+const (
+	Starting     = "Starting"
+	Running      = "Running"
+	ShuttingDown = "ShuttingDown"
+	Stopped      = "Stopped"
+	Restarting   = "Restarting"
+	Errored      = "Errored"
+)
+
+func (subprocess *Subprocess) Start(ctx context.Context) error {
+	var cancelCtx context.Context
+	cancelCtx, subprocess.cancel = context.WithCancel(ctx)
+
+	for k, v := range subprocess.config.EnvironmentVariables {
+		joined := fmt.Sprintf("%v=%v", k, v)
+		subprocess.envVars = append(subprocess.envVars, joined)
+	}
+
+	go func() {
+		subprocess.run(cancelCtx) // will block for lifetime of process
+		close(subprocess.shutdownSignal)
+	}()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (subprocess *Subprocess) Shutdown(ctx context.Context) error {
+	subprocess.cancel()
+	timeout := *subprocess.config.ShutdownTimeout
```

Comment:
Nit: not obvious from here that ShutdownTimeout cannot be nil, so looks a bit dangerous.

Reply:
Updated to use const to allow for unexpected api usage.